File: Data Structures & Algorithms/validate-parentheses/submission-3.cpp
class Solution {
public:
    bool isValid(string s) {
        bool res = true;
        stack<int> stk;

        for(int i=0;i<s.size();i++) {
            if(s[i] == '(' || s[i] == '{' || s[i] == '[') {
                stk.push(s[i]);
            }
            else {
                if(stk.size() == 0) {
                    res = false;
                    break;
                }
                char top = stk.top();
                if(s[i] == ')' && top != '(') {
                    res = false;
                    break;
                }
                else if(s[i] == '}' && top != '{') {
                    res = false;
                    break;
                }
                else if(s[i] == ']' && top != '[') {
                    res = false;
                    break;
                }
                stk.pop();
            }
        }
        if(stk.size() != 0){
            res = false;
        }

        return res;
    }
};

File: Data Structures & Algorithms/longest-common-subsequence/submission-3.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        #Bottom-Up; TC: O(n * m); SC: O(n * m)
        dp = [[0] * (len(text2) + 1) for _ in range(len(text1) + 1)]

        for i in range(len(text1) - 1, -1, -1):
            for j in range(len(text2) - 1, -1, -1):
                if(text1[i] == text2[j]):
                    dp[i][j] = 1 + dp[i + 1][j + 1]
                else:
                    dp[i][j] = max(dp[i + 1][j], dp[i][j + 1], dp[i + 1][j + 1])
        
        return dp[0][0]

        #Top - Down ; TC: O(n * m); SC: O(n * m)
        # dp = {}
        # def dfs(i: int, j: int) -> int:
        #     if(i == len(text1) or j == len(text2)):
        #         return 0
            
        #     if((i, j) in dp):
        #         return dp[(i, j)]

        #     if(text1[i] == text2[j]):
        #         res = 1 + dfs(i + 1,  j + 1)
        #     else:
        #         res = max(dfs(i + 1, j), dfs(i, j + 1), dfs(i + 1, j + 1))
            
        #     dp[(i, j)] = res
        #     return res

        # res = dfs(0, 0)
        
        # return res
